[RFC] win: build: make ci/build.ps1 work without Appveyor

janlazo opened the pull request:
Make the powershell build script not depend on Appveyor environment.

janlazo commented:
I guess it's cached or deleted during the build step so I can't copy the runtime files. I'll copy it at the end after running all the tests so I'm going to fix the `ci/build.ps1` to work with Visual Studio 2017 without running it.

For now, I'll default to MSVC if `%CONFIGURATION%` env var is undefined.

justinmk commented:
I don't follow. Did the change to `GetBinaryDep(...)` somehow cause MSVC build to fail? Will `build.ps1` be required for developers to build now?

janlazo commented:
It didn't fail but I don't know how to make cmake copy the nvim-qt runtime directory such that, in the `build/share/nvim-qt` is not created. I tried to add back  the same command prior to adding gui_shim.vim but changing `share/nvim/runtime` to `share/nvim-qt/runtime` so that nvim-qt can set its env var for its runtime directory and append a valid path to `rtp`.

Since I don't want to second-guess which directory nvim-qt runtime files are, I'm running `ci/build.ps1` on my machine without running Visual Studio 2017 GUI. I'm not even using its command prompt since cmake can find the Visual Studio folders anyway.

`ci/build.ps1` isn't required but I want to support running it on a regular cmd/powershell session without GUIs, same as in Appveyor.

janlazo commented:
cmake warnings are "transferred" to the next cmake invocation so the functional tests always fails on my machine. Not sure if this is cmake or msbuild.

justinmk commented:
this?

https://github.com/neovim/neovim/blob/e121b1dbe78cdf7ad46f493ca3a1cb83c190f719/ci/build.ps1#L124-L128

janlazo commented:
Yes. I'll deal with that later.

Debugging the following failures on my machine
```
 [  FAILED  ] 5 tests, listed below:
  [  FAILED  ] ...d/repo/git/neovim/test/functional\ex_cmds\write_spec.lua @ 34: :write &backupcopy=auto preserves symlinks
  ...d/repo/git/neovim/test/functional\ex_cmds\write_spec.lua:47: Expected objects to be the same.
  Passed in:
  (table) {
   *[1] = 'content0' }
  Expected:
  (table) {
   *[1] = 'content1' }

  stack traceback:
        ...d/repo/git/neovim/test/functional\ex_cmds\write_spec.lua:47: in function <...d/repo/git/neovim/test/functional\ex_cmds\write_spec.lua:34>

  [  FAILED  ] ...d/repo/git/neovim/test/functional\ex_cmds\write_spec.lua @ 68: :write appends FIFO file
  ...d/repo/git/neovim/test/functional\ex_cmds\write_spec.lua:82: test_fifo: No such file or directory

  stack traceback:
        ...d/repo/git/neovim/test/functional\ex_cmds\write_spec.lua:82: in function <...d/repo/git/neovim/test/functional\ex_cmds\write_spec.lua:68>

  [  FAILED  ] .../neovim/test/functional\legacy\011_autocommands_spec.lua @ 77: file reading, writing and bufnew and filter autocommands BufReadPre, BufReadPost (using gzip)
  .../neovim/test/functional\legacy\011_autocommands_spec.lua:91: Expected objects to be the same.
  Passed in:
  (string) '▼┬ì<├┐Z ♥Xtestfile +.I,*Q├êOS(I-.I├ï├îI├Ñ├è├ë├îKU0├ótLJNIMK├Å├ê├î├è├Ä├ë├ì├ï/(,*.)-+┬»┬¿┬¼┬é┬
¿0├ª┬¼├Ç♂ ┬¬┬ÜcJ┬ö9f♦├ì1'├è∟♂┬é├ªX↕e┬Ä┬í┬ü☻▲┬âR├│RP┬é¶ !qs&e☺  '
  Expected:
  (string) 'start of testfile
  line 2        Abcdefghijklmnopqrstuvwxyz
  line 3        xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
  line 4        Abcdefghijklmnopqrstuvwxyz
  line 5        xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
  line 6        Abcdefghijklmnopqrstuvwxyz
  line 7        xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
  line 8        Abcdefghijklmnopqrstuvwxyz
  line 9        xxxxxxxxxxxxxxxxxxxxxxxxxxxxxx
  line 10 Abcdefghijklmnopqrstuvwxyz
  end of testfile'

  stack traceback:
        .../neovim/test/functional\legacy\011_autocommands_spec.lua:91: in function <.../neovim/test/functional\legacy\011_autocommands_spec.lua:77>

  [  FAILED  ] .../neovim/test/functional\legacy\011_autocommands_spec.lua @ 98: file reading, writing and bufnew and filter autocommands FileReadPre, FileReadPost
  .../neovim/test/functional\legacy\011_autocommands_spec.lua:32: Expected objects to be the same.
  Passed in:
  (table) {
    [access] = 1526676621
    [change] = 1526676621
    [dev] = 2
    [gid] = 0
    [ino] = 0
    [mode] = 'file'
    [modification] = 1526676621
    [nlink] = 1
    [permissions] = 'rw-rw-rw-'
    [rdev] = 2
    [size] = 357
    [uid] = 0 }
  Expected:
  (nil)

  stack traceback:
        .../neovim/test/functional\legacy\011_autocommands_spec.lua:32: in function 'prepare_gz_file'
        .../neovim/test/functional\legacy\011_autocommands_spec.lua:99: in function <.../neovim/test/functional\legacy\011_autocommands_spec.lua:98>

  [  FAILED  ] ...d/repo/git/neovim/test/functional\legacy\delete_spec.lua @ 44: Test for delete() symlink delete
  ...d/repo/git/neovim/test/functional\legacy\delete_spec.lua:56: Expected objects to be the same.
  Passed in:
  (number) -1
  Expected:
  (number) 0

  stack traceback:
        ...d/repo/git/neovim/test/functional\legacy\delete_spec.lua:56: in function <...d/repo/git/neovim/test/functional\legacy\delete_spec.lua:44>
```

janlazo commented:
Is there a convenient way to query for admin privs other than `os.execute('mklink tmplink tmpfile')`?

janlazo commented:
@justinmk Does cmake still have an issue with sh/bash in `PATH`?

janlazo commented:
I copied only the shim because I don't know how to copy the entire runtime/ for nvim-qt. Feedback is appreciated.

In the meantime, I'll work on more path checks and adding a mingw fallback if Visual Studio is not installed.

justinmk commented:
> Does cmake still have an issue with sh/bash in PATH ? What was the reason why it's removed again?

It was only a problem when building with mingw, it comes from f53c8258bed7354678e5e355efa2acc471deb11a. Only @equalsraf knows why. If it doesn't cause a problem, maybe it's not a problem anymore.

> I disabled some tests (ie. gzip, fifo) because they come with msysgit and I don't want to deal with them now.

Ok. But I'm confused, how is that related to the gui shim? Aren't those tests already passing on master?

janlazo commented:
> Aren't those tests already passing on master?

They were skipped because Windows doesnt have gzip and mkfifo. I have them in my PATH because I install Git in a different folder. sh is not a problem in MSVC build.

janlazo commented:
Having issues compiling LuaJIT because of sjlj exceptions. Not sure if it's caused by `-O2`

janlazo commented:
cmake requires `-DCMAKE_SH=CMAKE_SH-NOTFOUND -DCMAKE_MAKE_PROGRAM=C:/msys64/mingw$bits/bin/mingw32-make` to make `cmake -G` work with Mingw Makefiles.

Maybe it's faster to patch cmake script instead.

erw7 commented:
@janlazo  How about installing `mingw-w64-$arch-ninja` and using `cmake - G Ninja`?  It does not complain even if there is sh in the PATH.

janlazo commented:
@erw7 I'll test after going through all the affected dependencies.

janlazo commented:
@erw7 Are you getting any errors for incremental/restarted builds? I can do a full build now with an empty `.dep/`. I'll proceed with Ninja afterwards but keep the changes if the generator is `MinGW Makefiles`.

justinmk commented on test/functional/legacy/011_autocommands_spec.lua:
```diff
@@ -61,7 +62,7 @@ describe('file reading, writing and bufnew and filter autocommands', function()
   if not has_gzip() then
     pending('skipped (missing `gzip` utility)', function() end)
-  else
+  elseif not iswin() then
```
instead change the condition on line 62 to:

    if iswin() or not has_gzip() then

justinmk commented on ci/build.ps1:
```diff
@@ -35,61 +43,83 @@ if ($compiler -eq 'MINGW') {
-
-cmd /c npm.cmd install -g neovim ; exitIfFailed
-where.exe neovim-node-host.cmd ; exitIfFailed
+# Perform sanity checks after installation
```
this comment is out of place here , better to remove it.

justinmk commented on ci/build.ps1:
```diff
@@ -35,61 +43,83 @@ if ($compiler -eq 'MINGW') {
+  $env:PATH = "C:\Python27;$env:PATH"
+}
+if (where.exe python.exe 2> $null) {
+  python.exe  -m pip install neovim ; exitIfFailed
```
the "Disambiguate python3" step should be done before calling python. Otherwise it might work by accident, but that is too subtle, better to be explicit.

equalsraf commented:
> It was only a problem when building with mingw, it comes from f53c825. Only @equalsraf knows why. If it doesn't cause a problem, maybe it's not a problem anymore.

This is a problem when using cmake with Mingw Makefiles, its an on and off thing, so watch out (refs [cmake-ml](https://cmake.org/pipermail/cmake/2016-February/062722.html) and [SO](https://stackoverflow.com/questions/41333215/mingw-sh-exe-must-not-be-in-your-path)).

The following are my [notes](https://gist.github.com/equalsraf/8710cd7324df0f67d047a99418356317) from last year, so yes I'm paraphrasing myself and could be wrong.

This is the error you get, cmake refuses to proceed if sh.exe is in PATH when using "MinGW Makefiles"

```
CMake Error at C:/msys64/mingw64/share/cmake-3.7/Modules/CMakeMinGWFindMake.cmake:12 (message):
  sh.exe was found in your PATH, here:
  C:/projects/neovim/.deps/build/src/libtermkey-build/"CMAKE_SH-NOTFOUND"
  For MinGW make to work correctly sh.exe must NOT be in your path.
  Run cmake from a shell that does not have sh.exe in your PATH.
  If you want to use a UNIX shell, then use MSYS Makefiles.
Call Stack (most recent call first):
```

If you retry it enough times it will actually work. But some of the third-party recipes have nested calls to cmake that lack the previous work around. You may be able to work around this using -DCMAKE_SH="CMAKE_SH-NOTFOUND" but this needs further study.

We work around this issue in appveyor builds by removing all path entries that have sh.exe (git, and msys2/usr/bin) while specifying some binary paths manually.

Finding a work around for this and patching third-party would probably simplify the appveyor PATH weirdness a good bit. But I keep getting contradictory results on this.

erw7 commented:
@janlazo I have not tried building `.dep` by Ninja. As I tried it, I needed to change [my branch](https://github.com/erw7/neovim/tree/feature-build-ninja-on-Windows) to succeed.

You can do cherry-pick if necessary. I wouldn't mind I create a new pull request.

janlazo commented on ci/build.ps1:
```diff
@@ -35,61 +43,83 @@ if ($compiler -eq 'MINGW') {
+  python.exe  -m pip install neovim ; exitIfFailed
+  python.exe  -c "import neovim; print(str(neovim))" ; exitIfFailed
+}
+if (Test-Path -PathType Container C:\Python35) {
```
Python/Ruby/Node only or every absolute path?

justinmk replied:
I don't know what you mean. We just need to know if our assumptions gold, rather than silently skipping CI.

janlazo commented:
@erw7 In your branch, why is mingw32-make required for Ninja? Isn't installing the ninja package on MinGW enough?

janlazo commented:
@equalsraf Setting `CMAKE_MAKE_PROGRAM` to absolute path of mingw32-make and passing this to all cmake calls along with `CMAKE_SH=CMAKE_SH-NOTFOUND` should be enough but why does cmake check for sh.exe? Is it to ensure that the generated Makefiles between MinGW and MSYS are different? As long as there's only 1 msys/mingw/cygwin environment included from `PATH`, there should be zero conflicts.

erw7 commented:
@janlazo  I modified to use cmake to build libuv. But builds of luajit must be done using Makefile. For that, `mingw32-make` is required.

justinmk commented on test/functional/ex_cmds/wviminfo_spec.lua:
```diff
@@ -36,7 +37,8 @@ describe(':wshada', function()
     -- sanity check
-    eq(text, io.open(shada_file):read())
+    local fp0 = io.open(shada_file)
+    eq(text, fp0:read())
```
`helpers.read_file()` does the open-read-close in one step. Or must the close be done at the end where you added it below?

janlazo replied:
I assumed that the order is important and I haven't used `:wshada` before. Does the test change if I use `helpers.read_file()`?

justinmk replied:
@janlazo good thought, but this was accidental. 

    eq(text, read_file(shada_file))

janlazo replied:
Can I do the same for the`fp:read()` after `:wshada`?

justinmk replied:
@janlazo I would leave it as-is since it's specifically getting only the first char.

equalsraf commented:
> but why does cmake check for sh.exe? 

No idea, it seems mingw32-make used to have issues with this, but maybe it does not anymore. The only references I found were these:

- https://cmake.org/pipermail/cmake/2016-February/062722.html
- https://cmake.org/pipermail/cmake/2016-June/063739.html

erw7 commented:
> but why does cmake check for sh.exe?

@janlazo , @equalsraf The second case is Japanese, but I found the following information.

-  https://stackoverflow.com/questions/24066265/force-mingw32-make-to-ignore-sh
- https://qiita.com/EmEpsilon/items/c1a0b5e60edfe71cb38a

There seems to be a case where `sh.exe` is used and an error occurs when there is a rule dependent on `cmd.exe`.

As it is more than 10 years ago, there is a possibility that the he himself may not remember, but it may be that @billhoffman who made this change knows something.

https://github.com/Kitware/CMake/commit/14cbc9571a7986617877fed0d2f04899d981d4ea#diff-26b18319c4c3e72f3fcf5a4d35ee0d55

janlazo commented:
@erw7 Can you make a PR for the Ninja build so that I can rebase on top of it?

erw7 commented:
@janlazo I am sorry, but can you also merge the changes I made today https://github.com/neovim/neovim/commit/fefb67fa885873c40577fc2bdc048204f8896a7a, https://github.com/neovim/neovim/commit/655e3167eec6196c3e480b360cfb0b93094a3359, https://github.com/neovim/neovim/commit/4feae966b96e8dd59c5fbb9a96849447ac3285c4, https://github.com/neovim/neovim/commit/091e81d650af49219fd05dfc420f8c6c803f62e7?

justinmk commented:
This is getting too big (lots of delicate changes that could have subtle effects). can someone break out the changes from @erw7 to a separate PR, if they are ready?

erw7 commented:
I have created a new [PR](https://github.com/neovim/neovim/pull/8434) that contain only Ninja build on Windows.

@janlazo I have created a [patch](https://github.com/neovim/neovim/commit/5152e29405c81eafe9594cb68606d0b5debe11a9.patch) that revert the merge of my branch, please use it if you want.

janlazo commented:
@erw7 Thanks for the patch but I'd rather wait for your PR to be merged first.

justinmk commented:
> I intend to keep support for MinGW Makefiles generator

If you find it personally useful then OK, otherwise I think it's fine to require Ninja on Windows: MSVC cmake support already requires it, and installing ninja in mingw is a single command.

janlazo commented:
Intent is to maintain support for MinGW Makefiles until the next (point) release so developers have time to adjust to the new build. I'll redo the commits so that it's easier to revert.

janlazo committed:
win/build: set CMAKE_SH,CMAKE_MAKE_PROGRAM

`MinGW Makefiles` build needs it if sh.exe is in PATH.

janlazo committed:
ci/win: validate paths/vars and add fallbacks

Build with MSVC if $env:CONFIGURATION is unset.
Add codepath for Ninja generator on MinGW build.
Check git, python, ruby paths from Appveyor to not break $env:PATH.
Run codecov and oldtests if mingw is installed
Stop the build if any cmdlet fails or a program returns a terminating error.
Force all (script) local variables to be set before using them.

janlazo committed:
win/build: remove -O flags for Luajit on MinGW

gcc enables sjlj exceptions but they're not native on Windows.

janlazo commented:
Is there a mingw compiler that uses seh exception for LuaJIT? I don't know how to force it for gcc so I'm leaving this PR as **RFC** for now.

justinmk commented:
I don't know how to answer that, maybe @jszakmeister @jamessan or some other build expert can help.

janlazo commented:
[makedeps.bat](https://github.com/neovim/neovim/commit/90e3dd92209c02d2717739461d4a6c8ed9ad2f7c) was included to conveniently build `.deps/` so it will be deprecated since I'm getting tired of running everything after updating the build script.

justinmk commented on ci/build.ps1:
```diff
@@ -35,95 +45,131 @@ if ($compiler -eq 'MINGW') {
 
   # Build third-party dependencies
+  $mingwPackages = @('gcc', 'make', 'cmake', 'perl', 'diffutils', 'unibilium').ForEach({
+    @('mingw', 'w64', $arch, $_) -join '-'
```
The old code was more obvious. Is there a functional need for this change?

janlazo replied:
No. I was testing the pattern so I can conveniently add gcc, make, ninja. I prefer joining arrays than `'mingw-w64-' + $arch + '-' + $_` and all packages had the same naming pattern.

justinmk commented:
`build.ps1` is getting complicated, maybe it isn't worth trying to make it work outside of AppVeyor. Instead use a separate script, then we can potentially use _that_ from `ci/build.ps1` if it has a lot of duplicated ideas.  But I would guess it might end up not needing much code at all, e.g. `makedeps.bat` is trivial.

 What exactly do you use `build.ps1` for, other than building `.deps` ?

janlazo commented:
@justinmk I want a convenient script to run everything and `ci/build.ps1` has that use-case. I wasn't compiling nvim at all prior to this PR. Problem is the port from batchfile to powershell script is incomplete (for me) because it didn't take full advantage of powershell features.

`exitIfFailed` might be redundant now. There is `makedeps.bat` but I don't want to write a batchfile when a powershell function suffices.

janlazo committed:
ci/win: enable MSVC_64,MINGW_32 in PRs

janlazo commented:
I can't reproduce the error for xgettext on MSVC_64. It could be a path escaping issue with one of the arguments if it uses cmd.exe.

MINGW_32 is completely broken. Could be a segfault (overflow) because of the negative exit code. Strange that all of the dependencies are fine such that nvim can compile.
